### ISSUE TITLE
Add `file://` protocol to `self` links that currently don't have a protocol

### ISSUE THREAD

**Issue by gadomski** [stac-utils/pystac#1347]
xref https://github.com/radiantearth/stac-spec/pull/1281, but it seems likely that at some point in the future `self` links will be required to have a protocol (i.e. absolute filesystem paths won't be enough). We should update to **pystac** to write `file://` absolute paths to future-proof ourselves and help ecosystem adoption.

**Comment by jsignell** [comment]
Does this also imply that RELATIVE_PUBLISHED is no longer a valid catalog type?

**Comment by jsignell** [comment]
Oh ok so absolute in this context just means a url with a protocol. Not as in absolute vs relative path?

**Comment by gadomski** [comment]
> Oh ok so absolute in this context just means a url with a protocol. Not as in absolute vs relative path?

Yeah, I think that text from the best practices is now a bit confusing since a href with a protocol is probably always absolute ... I can't think of an example of a `protocol://`-style href that could be relative?

**Comment by jsignell** [comment]
I was thinking about this more in the context of the migrate function. I think the next version of pystac should be able to read URIs that have the `file:///` protocol, but I _think_ it should not write them that way until version 2.0. It just feels like too big of a change to put in a minor release.

**Comment by jsignell** [comment]
I'll open a new PR now taking just the reading bits from #1472
